revert: 4023228 ("let's exception not bubble")

pajod opened the pull request:
The report in https://github.com/benoitc/gunicorn/issues/2923 may have been not caused by a bug in Gunicorn or gevent in the first place. In any case, going from "this particular exception" to "each and every BaseException" was *way* excessive and is a likely explanation for https://github.com/benoitc/gunicorn/issues/3207

Context:
* https://github.com/benoitc/gunicorn/commit/40232284934c32939c0e4e78caad1987c3773e08#r144628336

Suggested changes:
* revert that patch
* reintroduce a more specific exception catching that mimics what the reverted patch tried to accomplish, while emitting a new warning.

Unfinished, but no longer needed after eventlet support was dropped:
* the SSLError from wrap_socket in the *eventlet* worker (afaict, can only be triggered by the client sending garbage or disconnecting while Gunicorn was overloaded) cannot be dealt with this way due to class inheritance stepping outside the huge try-except block. I have a [currently skipped test to reproduce this](https://github.com/pajod/gunicorn/blob/3370ad6be6ae7bf3be1a340d2fc4cfcbf9243975/tests/test_wrk.py#L346-L347) - the problem is over here: https://github.com/benoitc/gunicorn/blob/903792f152af6a27033d458020923cb2bcb11459/gunicorn/workers/geventlet.py#L153-L156
  * WIP: https://github.com/pajod/gunicorn/commit/f63030411982edf390149c8df5f10b8fdeaf8907
----
* Partially reverts: 40232284934c32939c0e4e78caad1987c3773e08
* Partially reverts: 0b10cbab1d6368fcab2d5a7b6fe359a6cecc81a7
* Fixes: https://github.com/benoitc/gunicorn/issues/3207
* Partially fixes: https://github.com/benoitc/gunicorn/issues/3410

pajod committed:
Revert "let's exception not bubble"

This reverts commit 40232284934c32939c0e4e78caad1987c3773e08.

We use sys.exit. On purpose.
We should therefore not be catching SystemExit.

pajod committed:
workaround: reintroduce gevent.Timeout handling

This is probably wrong. But less wrong than handling *all*
BaseException. Reports of this happening may be the result of
 some *other* async Timeout in the wsgi app bubbling through to us.

Gevent docs promise:
"[..] if *exception* is the literal ``False``, the timeout is still raised,
   but the context manager suppresses it, so
   the code outside the with-block won't see it."

pajod committed:
style: line break

benoitc commented:
@pajod PR must be provided over the master and don't assume that other PR may be added. There is no guarantee a separate PR will be acccepted, so  PR will be enough by itself. 

This PR sounds OK by itself. What could miss?

pajod committed:
Partially revert "unconditionally log request error"

There is not need to excessively produce high-severity logs, while
there are ways to reach that point from entirely benign network errors.

This partially reverts commit 0b10cbab1d6368fcab2d5a7b6fe359a6cecc81a7.

wholegrainloaf commented:
I found this PR through this issue: https://github.com/benoitc/gunicorn/issues/3207. @pajod are you still planning on merging this fix? Was it dropped in favor of a different solution? I'd love to see this fixed 😄
